pkg/innerring/processors/settlement/basic/collect.go

```diff
@@ -0,0 +1,96 @@
+package basic
+
+import (
+	"math/big"
+
+	"github.com/nspcc-dev/neofs-node/pkg/innerring/processors/settlement/common"
+	"github.com/nspcc-dev/neofs-node/pkg/morph/client/container/wrapper"
+	"go.uber.org/zap"
+)
+
+var (
+	bigGB   = big.NewInt(1 << 30)
+	bigZero = big.NewInt(0)
+	bigOne  = big.NewInt(1)
+)
+
+func (inc *IncomeSettlementContext) Collect() {
+	inc.mu.Lock()
+	defer inc.mu.Unlock()
+
+	// todo: save state of bank wallet
+
+	cachedRate, err := inc.rate.BasicRate()
+	if err != nil {
+		inc.log.Error("can't get basic income rate",
+			zap.String("error", err.Error()))
+
+		return
+	}
+
+	cnrEstimations, err := inc.estimations.Estimations(inc.epoch)
+	if err != nil {
+		inc.log.Error("can't fetch container size estimations",
+			zap.Uint64("epoch", inc.epoch))
+
+		return
+	}
+
+	for i := range cnrEstimations {
+		owner, err := inc.container.ContainerInfo(cnrEstimations[i].ContainerID)
+		if err != nil {
+			inc.log.Warn("can't fetch container info",
+				zap.Uint64("epoch", inc.epoch),
+				zap.Stringer("container_id", cnrEstimations[i].ContainerID))
+
+			continue
+		}
+
+		cnrNodes, err := inc.placement.ContainerNodes(inc.epoch, cnrEstimations[i].ContainerID)
+		if err != nil {
+			inc.log.Debug("can't fetch container info",
+				zap.Uint64("epoch", inc.epoch),
+				zap.Stringer("container_id", cnrEstimations[i].ContainerID))
+
+			continue
+		}
+
+		avg := inc.avgEstimation(cnrEstimations[i]) // average container size per node
+		total := calculateBasicSum(avg, cachedRate, len(cnrNodes))
+
+		inc.txTable.Transfer(&common.TransferTx{
+			From:   owner.Owner(),
+			To:     inc.bankOwner,
+			Amount: total,
+		})
+	}
+
+	common.TransferAssets(inc.exchange, inc.txTable)
+}
+
+// avgEstimation returns estimation value for single container. Right now it
+// simply calculates average of all announcements, however it can be smarter and
+// base result on reputation of announcers and clever math.
+func (inc *IncomeSettlementContext) avgEstimation(e *wrapper.Estimations) (avg uint64) {
+	for i := range e.Values {
+		avg += e.Values[i].Size
+	}
+
+	return avg / uint64(len(e.Values))
```

Comment:
Apparently a requirement is imposed on `EstimationFetcher.Estimations` that all `Estimation.Value` be non-empty. Please check that all implementations are following this.

Reply:
Thanks, I'll check this, missed possible 0-division.

Reply:
Fixed.